newapi/ooniapi/auth.py

```diff
@@ -0,0 +1,406 @@
+    port = conf["MAIL_PORT"]
+    mail_user = conf["MAIL_USERNAME"]
+    mail_password = conf["MAIL_PASSWORD"]
+    use_ssl = conf["MAIL_USE_SSL"]
+    log.debug(f"connecting to {smtphost}:{port} as {mail_user}")
+    SMTP = smtplib.SMTP_SSL if use_ssl else smtplib.SMTP
+    try:
+        with SMTP(host=smtphost, port=port) as s:
+            s.ehlo()
+            s.login(mail_user, mail_password)
+            s.send_message(msg)
+    except Exception as e:
+        log.error(e, exc_info=1)
+        raise
+
+
+def send_login_email(dest_addr, nick, token: str) -> None:
+    """Format and send a registration/login  email"""
+    src_addr = current_app.config["MAIL_SOURCE_ADDRESS"]
+    baseurl = current_app.config["BASE_URL"]
+    url = urljoin(baseurl, f"/api/v1/user_login?k={token}")
+
+    msg = EmailMessage()
+    msg["Subject"] = "OONI Account activation"
+    msg["From"] = src_addr
+    msg["To"] = (dest_addr,)
+
+    txt = f"""Welcome to OONI, {nick}.
+
+    Please login by following {url}
+
+    The link can be used on multiple devices and will expire in 24 hours.
+    """
+    msg.set_content(txt)
+    html = f"""\
+<html>
+  <head></head>
+  <body>
+    <p>Welcome to OONI, {nick}</p>
+    <p>
+        <a href="{url}">Please login here </a>
+    </p>
+    <p>The link can be used on multiple devices and will expire in 24 hours.</p>
+  </body>
+</html>
+"""
+    msg.add_alternative(html, subtype="html")
+    _send_email(dest_addr, msg)
+
+
+@metrics.timer("user_register")
+@auth_blueprint.route("/api/v1/user_register", methods=["POST"])
+def user_register():
+    """Auth Services: start email-based user registration
+    ---
+    parameters:
+      - in: body
+        name: register data
+        description: Registration data as HTML form or JSON
+        required: true
+        schema:
+          type: object
+          properties:
+            nickname:
+              type: string
+            email_address:
+              type: string
+    responses:
+      200:
+        description: Confirmation
+    """
+    log = current_app.logger
+    req = request.json if request.is_json else request.form
+    nick = req.get("nickname", "").strip()
+    # Accept all alphanum including unicode and whitespaces
+    if not nick.replace(" ", "").isalnum():
+        return jerror("Invalid user name")
+    if len(nick) < 3:
+        return jerror("User name is too short")
+    if len(nick) > 50:
+        return jerror("User name is too long")
+
+    email_address = req.get("email_address", "").strip().lower()
+    if not nick or not email_address:
+        return jerror("Invalid request")
+    if EMAIL_RE.fullmatch(email_address) is None:
+        return jerror("Invalid email address")
+
+    account_id = hash_email_address(email_address)
+    now = datetime.utcnow()
+    expiration = now + timedelta(days=1)
+    # On the backend side the registration is stateless
+    payload = {
+        "nbf": now,
+        "exp": expiration,
```

Comment:
I believe you have to convert these `datetime` objects into an int of seconds since epoch.

I tried running:
```
jwt.encode({'a': datetime.utcnow()}, "antani")
```

and I get
````
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/art/.virtualenvs/ooni-api/lib/python3.7/site-packages/jwt/api_jwt.py", line 60, in encode
    payload, separators=(",", ":"), cls=json_encoder
  File "/Users/art/.pyenv/versions/3.7.3/lib/python3.7/json/__init__.py", line 238, in dumps
    **kw).encode(obj)
  File "/Users/art/.pyenv/versions/3.7.3/lib/python3.7/json/encoder.py", line 199, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/Users/art/.pyenv/versions/3.7.3/lib/python3.7/json/encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
  File "/Users/art/.pyenv/versions/3.7.3/lib/python3.7/json/encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type datetime is not JSON serializable
```

Reply:
The library handles the special keys by itself:
`jwt.encode({'nbf': datetime.utcnow()}, "antani")`